### PR TITLE
[video][android][ios] Add video caching

### PR DESCRIPTION
**Pull request by behenate**
# Why

Adds video caching for regular video files for Android and iOS. On Android HLS videos also work.

# How

On Android used `SimpleCache` to implement the cache. On iOS used a `AVAssetResourceLoaderDelegate` to capture the requests, fulfill them and cache the responses.

The iOS structure
- `ResourceLoaderDelegate` - implements AVAssetResourceLoaderDelegate, URLSessionDelegate - recieves requests from the AVPlayer and fulfills them using an URL session. When receiving a request checks if the data is available in the CachedResource and fully/partially fulfills it from the cache.
- `CachedResource` - Meant to represent a whole of a cached video (metadata + video data).
- `CachableRequest `- Holds a loadingRequest and associated variables for each loadingRequest received from the AVPlayer. Holds the data received from the server until the loadingRequest is fulfilled.
- `MediaInfo` - Holds information necessary to play the video from a file, including the mimeType, headers received from the server response and loadedDataRanges - ranges of data that are already available in the file.

# Test Plan

Tested in BareExpo on Android and iOS

### CODE REVIEW

File: packages/expo-video/CHANGELOG.md [hunk 1]
```diff
@@ -9,10 +9,11 @@
 ### 🎉 New features
 
 - [iOS] Add 'allowsExternalPlayback' property to control the AirPlay to stream video or mirror the screen. ([#30436](https://github.com/expo/expo/pull/30436) by [@adsalihac](https://github.com/adsalihac))
-- [iOS] Fixed `player.currentTime` being `NaN` when source is not provided and `player.duration` being `NaN` inside the hook callback when the source is updated. ([#31011](https://github.com/expo/expo/pull/31011) by [@AlirezaHadjar](https://github.com/AlirezaHadjar))
+- [Android][iOS] Add caching functionality. ([#31781](https://github.com/expo/expo/pull/31781) by [@behenate](https://github.com/behenate))
```
Comment (lukmccall) [inline packages/expo-video/CHANGELOG.md:12]:
```suggestion
- Add caching functionality. ([#31781](https://github.com/expo/expo/pull/31781) by [@behenate](https://github.com/behenate))
```

File: packages/expo-video/android/src/main/java/expo/modules/video/VideoCache.kt [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    }
+  private val databaseProvider: DatabaseProvider = StandaloneDatabaseProvider(context)
+  private var cacheEvictor: LeastRecentlyUsedCacheEvictor
+  private val sharedPreferences = context.getSharedPreferences(SHARED_PREFERENCES_NAME, Context.MODE_PRIVATE)
+  var instance: SimpleCache
+
+  init {
+    cacheEvictor = LeastRecentlyUsedCacheEvictor(getMaxCacheSize())
+    instance = SimpleCache(getCacheDir(), cacheEvictor, databaseProvider)
+  }
```
Comment (lukmccall) [inline packages/expo-video/android/src/main/java/expo/modules/video/VideoCache.kt:32]:
```suggestion
  private val sharedPreferences = context.getSharedPreferences(SHARED_PREFERENCES_NAME, Context.MODE_PRIVATE)
  private var cacheEvictor = LeastRecentlyUsedCacheEvictor(getMaxCacheSize())
  var instance = SimpleCache(getCacheDir(), cacheEvictor, databaseProvider)
```

File: packages/expo-video/android/src/main/java/expo/modules/video/VideoCache.kt [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+  private fun getFileSize(file: File): Long {
+    var size: Long = 0
+    if (file.isDirectory) {
+      file.listFiles()?.forEach {
+        size += getFileSize(it)
+      }
+    } else {
+      size = file.length()
+    }
+    return size
```
Comment (lukmccall) [inline packages/expo-video/android/src/main/java/expo/modules/video/VideoCache.kt:105]:
```suggestion
    if (file.isDirectory) {
    	file.listFiles().fold(0L) { sum, element -> sum + getFileSize(element) }
    }
    
    return file.lenght()
```
Something like this should work

File: packages/expo-video/ios/VideoPlayerItem.swift [hunk 4]
```diff
@@ -1,11 +1,25 @@
 // Copyright 2024-present 650 Industries. All rights reserved.
 
 import AVFoundation
-
-class VideoPlayerItem: AVPlayerItem {
+import ExpoModulesCore
```
Comment (tsapeta) [inline packages/expo-video/ios/VideoPlayerItem.swift:4]:
```suggestion
import ExpoModulesCore

```

File: packages/expo-video/ios/VideoPlayerItem.swift [hunk 5]
```diff
@@ -1,11 +1,25 @@
...
+    let canCache = Self.canCache(videoSource: videoSource)
+    let shouldCache = videoSource.useCaching && canCache
+
+    if !canCache && videoSource.useCaching {
+      log.warn("Provided source with uri: \(videoSource.uri?.absoluteString ?? "null") cannot be cached. Caching will be disabled")
+    }
+    super.init(url: url, useCaching: shouldCache, avUrlAssetOptions: avUrlAssetOptions)
+  }
+
+  private static func canCache(videoSource: VideoSource) -> Bool {
```
Comment (tsapeta) [inline packages/expo-video/ios/VideoPlayerItem.swift:19]: It is private and doesn't depend on the class, so let's make it global

File: packages/expo-video/src/VideoModule.ts [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+ * @platform android
+ * @platform ios
+ */
+export function isPictureInPictureSupported(): boolean {
+  return NativeVideoModule.isPictureInPictureSupported();
+}
+
+/**
+ * Cleans all video cache.
+ * > This function can be called only if there are no existing `VideoPlayer` instances.
```
Comment (tsapeta) [inline packages/expo-video/src/VideoModule.ts:16]: What happens when there is an instance? Throws an error?

File: packages/expo-video/src/VideoModule.ts [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+ * @platform ios
+ */
+export function cleanVideoCacheAsync(): Promise<void> {
+  return NativeVideoModule.cleanVideoCacheAsync();
+}
+
+/**
+ * Sets desired video cache size in bytes. The default video cache size is 1GB. Value set by this function is persistent.
+ * The cache size is not guaranteed to be exact and the actual cache size may be slightly larger. The cache is evicted on a least-recently-used basis.
+ * > This function can be called only if there are no existing `VideoPlayer` instances.
```
Comment (tsapeta) [inline packages/expo-video/src/VideoModule.ts:29]: Same as above.

File: docs/pages/versions/unversioned/sdk/video.mdx [hunk 8]
```diff
@@ -249,6 +249,24 @@ const styles = StyleSheet.create({
 
 </SnackInline>
 
+## Caching videos (experimental)
+
+> **Info** Caching was introduced during the final stages of the beta phase for `expo-video`, and it has not been thoroughly tested by the community yet.
+> If you encounter any issues, please report them at the [expo/expo](https://github.com/expo/expo) repository.
```
Comment (tsapeta) [inline docs/pages/versions/unversioned/sdk/video.mdx:255]: Long time has passed since then, `expo-video` is no longer in beta. Can you rephrase it a bit?

File: docs/pages/versions/unversioned/sdk/video.mdx [hunk 9]
```diff
@@ -249,6 +249,24 @@ const styles = StyleSheet.create({
...
+
+If your app frequently replays the same video, caching can be utilized to minimize network usage and enhance user experience, albeit at the cost of increased device storage usage.
+`expo-video` supports video caching on `Android` and `iOS` platforms. This feature can be activated by setting the [`useCache`](#videosource) property of a [`VideoSource`](#videosource) object to `true`.
```
Comment (tsapeta) [inline docs/pages/versions/unversioned/sdk/video.mdx:258]:
```suggestion
`expo-video` supports video caching on `Android` and `iOS` platforms. This feature can be activated by setting the [`useCaching`](#videosource) property of a [`VideoSource`](#videosource) object to `true`.
```

File: packages/expo-video/src/VideoPlayer.types.ts [hunk 10]
```diff
@@ -229,6 +229,17 @@ export type VideoSource =
        * @platform ios
        */
       headers?: Record<string, string>;
+
+      /**
+       * @experimental
+       * Specifies whether the player should use caching for the video.
+       * > Due to platform limitations caching HLS video sources on iOS is not supported.
+       * > Caching DRM protected videos is not supported.
```
Comment (tsapeta) [inline packages/expo-video/src/VideoPlayer.types.ts:237]:
```suggestion
       * > Caching DRM-protected videos is not supported.
```

File: docs/pages/versions/unversioned/sdk/video.mdx [hunk 11]
```diff
@@ -249,6 +249,24 @@ const styles = StyleSheet.create({
...
+The cache is persistent and will be cleared on a least-recently-used basis once the preferred size is exceeded. Furthermore, the system can clear the cache due to low storage availability, so it's not advisable to depend on the cache to store critical data.
+The cache functions offline. If a portion or the entirety of a video is cached, it can be played from the cache even when the device is offline until the cached data is exhausted.
+
+> The cache cannot be used with HLS video sources and DRM-protected videos.
+
+### Managing the Cache
+
+- The preferred cache size in bytes can be defined using the [`setVideoCacheSizeAsync`](#videosetvideocachesizeasyncsizebytes) function. The default cache size is `1GB`.
+- The [`getCurrentVideoCacheSize`](#videogetcurrentvideocachesize) can be used to get the current storage occupied by the cache in bytes.
+- All cached videos can be cleared using the [`cleanVideoCacheAsync`](#videocleanvideocacheasync) function.
```
Comment (tsapeta) [inline docs/pages/versions/unversioned/sdk/video.mdx:268]: Here you use both words, `clear` and `clean` 😄 Let's keep things consistent.

**Review by tsapeta — APPROVED** [review]

File: packages/expo-video/CHANGELOG.md (hunk 1 above)
Comment (behenate) [reply packages/expo-video/CHANGELOG.md:12]: I think we should keep the labels, it's not available for web.

File: packages/expo-video/src/VideoModule.ts (hunk 6 above)
Comment (behenate) [reply packages/expo-video/src/VideoModule.ts:16]: Yep throws `VideoCacheException("Cannot clear cache while there are active players")`

File: docs/pages/versions/unversioned/sdk/video.mdx [hunk 12]
```diff
@@ -294,6 +294,21 @@ import { createVideoPlayer } from 'expo-video';
 const player = createVideoPlayer(videoSource);
 ```
 
+### Caching videos
+
+If your app frequently replays the same video, caching can be utilized to minimize network usage and enhance user experience, albeit at the cost of increased device storage usage.
+`expo-video` supports video caching on `Android` and `iOS` platforms. This feature can be activated by setting the [`useCaching`](#videosource) property of a [`VideoSource`](#videosource) object to `true`.
+The cache is persistent and will be cleared on a least-recently-used basis once the preferred size is exceeded. Furthermore, the system can clear the cache due to low storage availability, so it's not advisable to depend on the cache to store critical data.
+The cache functions offline. If a portion or the entirety of a video is cached, it can be played from the cache even when the device is offline until the cached data is exhausted.
```
Comment (amandeepmittal) [inline docs/pages/versions/unversioned/sdk/video.mdx:302]:
```suggestion
If your app frequently replays the same video, caching can be utilized to minimize network usage and enhance user experience, albeit at the cost of increased device storage usage. `expo-video` supports video caching on `Android` and `iOS` platforms. This feature can be activated by setting the [`useCaching`](#videosource) property of a [`VideoSource`](#videosource) object to `true`.

The cache is persistent and will be cleared on a least-recently-used basis once the preferred size is exceeded. Furthermore, the system can clear the cache due to low storage availability, so it's not advisable to depend on the cache to store critical data.

The cache functions offline. If a portion or the entirety of a video is cached, it can be played from the cache even when the device is offline until the cached data is exhausted.
```

Let's break this down into multiple paragraphs because right it is one big paragraph and there are different pieces of information discussed.

File: docs/pages/versions/unversioned/sdk/video.mdx [hunk 13]
```diff
@@ -294,6 +294,21 @@ import { createVideoPlayer } from 'expo-video';
...
+
+> The cache cannot be used with HLS video sources and DRM-protected videos.
+
+### Managing the Cache
```
Comment (amandeepmittal) [inline docs/pages/versions/unversioned/sdk/video.mdx:306]:
```suggestion
### Managing the cache
```

Let's stick to using sentence cases for section titles as per the writing style guide: https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md#headings

File: docs/pages/versions/unversioned/sdk/video.mdx [hunk 14]
```diff
@@ -294,6 +294,21 @@ import { createVideoPlayer } from 'expo-video';
...
+
+- The preferred cache size in bytes can be defined using the [`setVideoCacheSizeAsync`](#videosetvideocachesizeasyncsizebytes) function. The default cache size is `1GB`.
```
Comment (amandeepmittal) [inline docs/pages/versions/unversioned/sdk/video.mdx:308]:
```suggestion
- The preferred cache size in bytes can be defined using the [`setVideoCacheSizeAsync`](#videosetvideocachesizeasyncsizebytes) function. The default cache size is 1GB.
```

I don't think we need to apply code syntax inline highlight to the unit here.

File: docs/pages/versions/unversioned/sdk/video.mdx [hunk 15]
```diff
@@ -294,6 +294,21 @@ import { createVideoPlayer } from 'expo-video';
...
 ```
 
+### Caching videos
+
+If your app frequently replays the same video, caching can be utilized to minimize network usage and enhance user experience, albeit at the cost of increased device storage usage.
+`expo-video` supports video caching on `Android` and `iOS` platforms. This feature can be activated by setting the [`useCaching`](#videosource) property of a [`VideoSource`](#videosource) object to `true`.
+The cache is persistent and will be cleared on a least-recently-used basis once the preferred size is exceeded. Furthermore, the system can clear the cache due to low storage availability, so it's not advisable to depend on the cache to store critical data.
+The cache functions offline. If a portion or the entirety of a video is cached, it can be played from the cache even when the device is offline until the cached data is exhausted.
+
+> The cache cannot be used with HLS video sources and DRM-protected videos.
```
Comment (amandeepmittal) [inline docs/pages/versions/unversioned/sdk/video.mdx:304]: Based on the  following note below: 
CleanShot 2025-01-06 at 13 50 28@2x

I think we should keep the same verbiage here as well. Right now, it is confusing to see the same message in two places, with the difference being the iOS platform.

I think it'll be good to say here as well that: "Due to platform limitations, the cache cannot be used with HSL video sources on iOS. Caching DRM-protected videos is not supported on Android and iOS."

If this is technically right, can we also update the "Caching DRM-protected videos..." note in the type definition to mention that it is not supported on any platform?

File: packages/expo-video/src/VideoPlayer.types.ts [hunk 16]
```diff
@@ -282,6 +282,17 @@ export type VideoSource =
        * @platform ios
        */
       headers?: Record<string, string>;
+
+      /**
+       * @experimental
+       * Specifies whether the player should use caching for the video.
+       * > Due to platform limitations caching HLS video sources on iOS is not supported.
+       * > Caching DRM-protected videos is not supported.
```
Comment (amandeepmittal) [inline packages/expo-video/src/VideoPlayer.types.ts:290]:
```suggestion
       * > Caching DRM-protected videos is not supported on Android and iOS.
```

Based on the comment and context I left in the Expo Video API reference page earlier.

**Review by lukmccall — APPROVED** [review]

File: packages/expo-video/android/src/main/java/expo/modules/video/VideoCache.kt [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+    // Creates a new cache directory to avoid conflicts while removing the old cache
+    val oldCacheDirectory = getCacheDir()
+    val oldCache = instance
+    val newCacheName = generateCacheDirName()
+
+    sharedPreferences.edit().putString(VIDEO_CACHE_DIR_KEY, newCacheName).apply()
+    instance = SimpleCache(getCacheDir(), cacheEvictor, databaseProvider)
+    oldCache.release()
+    deleteRecursive(oldCacheDirectory)
+  }
```
[the remark this answers is not in the document]
Comment (behenate) [reply packages/expo-video/android/src/main/java/expo/modules/video/VideoCache.kt:85]: It doesn't for now. The first line asserts that there are no active players, I did it this way, because it seemed like handling the cache being deleted during playback would be too complex. We stated in the docs that there can be no active players when clearing the cache or changing its size.
